src/Language/Fixpoint/Solver/EnvironmentReduction.hs

```diff
@@ -568,40 +571,54 @@ mergeDuplicatedBindings xs =
 -- | Inlines some of the bindings whose symbol satisfies a given predicate.
 --
 -- Only works if the bindings don't form cycles.
-substBindings
-  :: Lens' v SortedReft
+substBindingsSimplifyingWith
+  :: (Expr -> Expr)
+  -> Lens' v SortedReft
   -> (Symbol -> Bool)
   -> HashMap Symbol v
   -> HashMap Symbol v
-substBindings vLens p env =
+substBindingsSimplifyingWith simplifier vLens p env =
     -- Circular program here. This should terminate as long as the
     -- bindings introduced by ANF don't form cycles.
-    let env' = HashMap.map (vLens %~ inlineInSortedReft (srLookup filteredEnv)) env
+    let env' = HashMap.map (vLens %~ sortedReftSimplifier . inlineInSortedReft (srLookup filteredEnv)) env
         filteredEnv = HashMap.filterWithKey (\sym _v -> p sym) env'
      in env'
   where
     srLookup env' sym = view vLens <$> HashMap.lookup sym env'
+    sortedReftSimplifier (RR s r) = RR s (reft (reftBind r) (simplifier (reftPred r)))
```

Comment:
I think that you intended to remove the `simplify` call from `inlineInExpr` which would be redundant now, and update the call sites of `inlineInExpr` and `inlineInSortedReft`.

Reply:
Yes, of course.  I think I confused myself earlier!